### DISCUSSION TITLE
Where is the code (of BucketingSampler) that splits the cuts up by rank when doing distributed training?

### DISCUSSION THREAD

**Discussion by RuABraun** [lhotse-speech/lhotse#1122] [Q&A] [answered]
I can't find it. I assume it's in the sampler code somewhere since that holds the cuts.

**Comment by pzelasko** [comment, marked as answer]
It's in the base `CutSampler` class https://github.com/lhotse-speech/lhotse/blob/c80fc0774de06b5f0064888a0774277b1727ef3f/lhotse/dataset/sampling/base.py#L275

**Reply by RuABraun** [reply]
I see i see. I skimmed the file but didnt notice, should have searched for self.rank or distributed. thanks!

**Reply by lifeiteng** [reply]
@pzelasko I don't understand the "trick" to ensure equal number of batches for each distributed worker.
As all random seed is fixed, assume batch sequence is 4(batch_size), 6, 7, 8, we have 4 GPUs, so
GPU0 batches = [4(batch_size), 6, 7, 8] -> batches[0] = 4
GPU1 batches = [4(batch_size), 6, 7, 8]  -> batches[1] = 6
GPU2 batches = [4(batch_size), 6, 7, 8] -> batches[2] = 7
GPU3 batches = [4(batch_size), 6, 7, 8] -> batches[3] = 8

This does not meet our expectations. What is wrong?

**Reply by pzelasko** [reply]
The number of batches per epoch is equal ensuring the GPUs have the same number of steps in the train/dev loop, but the batch size on each GPU is not necessarily equal. Does that make sense?

**Reply by lifeiteng** [reply]
Oh, I see, It's `The number of batches per epoch` not `The batch_size of per batch on  distributed  GPU`

It seems we didn't feed world_size & rank correctly, Am I right? BucketingSampler.world_size is always equal 1.
```
class BucketingSampler(CutSampler):
        super().__init__(
            drop_last=drop_last,
            world_size=1, 
            rank=0,
            seed=seed,
        )

class CutSampler(Sampler, Dillable):
     self._maybe_init_distributed(world_size=world_size, rank=rank)

    def _maybe_init_distributed(self, world_size: Optional[int], rank: Optional[int]):
        if world_size is not None:
            assert world_size >= 1
        if rank is not None:
            assert rank >= 0

        # Order of precedence:
        # 1. When world size or rank are explicitly provided, we will use them.
        # 2. Next, check WORLD_SIZE and RANK env variables; yes? use them.
        # 3. Next, check if torch.distributed is initialized and has them set; yes? use them.
        # 4. If none of those are available, rank=0 and world_size=1.
        if "WORLD_SIZE" in os.environ and "RANK" in os.environ:
            # If deepspeed launcher is being used, it will set the env variables automatically.
            self.world_size = ifnone(world_size, int(os.environ["WORLD_SIZE"]))
            self.rank = ifnone(rank, int(os.environ["RANK"]))
        elif dist.is_available() and dist.is_initialized():
            self.world_size = ifnone(world_size, dist.get_world_size())
            self.rank = ifnone(rank, dist.get_rank())
        else:
            self.world_size = ifnone(world_size, 1)
            self.rank = ifnone(rank, 0)
        assert self.rank < self.world_size
```

**Reply by pzelasko** [reply]
BucketingSampler has a specific implementation that delegates the per node splitting to internal SimpleCutSampler objects, so it has world size set to 1 but the internal samplers don’t. 

With BucketingSampler it’s actually possible each node sees roughly the same batch sizes as the internal bucket selection RNG would be likely initialized with the same seed on each rank. The batch size is similar but not exactly the same because of cut duration variation (eg sometimes 21, sometimes 23 cuts will hit max duration). However this synchronization will usually only last for a number of steps at the start of training and diverge over time, as dataloader workers have different computation time and put batches in FIFO buffer, and eventually they will go out of order.

**Reply by lifeiteng** [reply]
BucketingSampler needs `x RealWordSize`(as the `world_size==1`) steps to reach the end of Epoch, which is different with DynamicBucketingSampler. Does this meet expectation?

**Reply by pzelasko** [reply]
Are you sure? See this snippet which passes rank and world size to internal samplers if defined (and if not, they auto detect)

https://github.com/lhotse-speech/lhotse/blob/d3106cf374df5dc33b16a5fcdd2ecb87c0cd0359/lhotse/dataset/sampling/bucketing.py#L99

**Reply by lifeiteng** [reply]
I roughly figured it out, thanks.

**Reply by RuABraun** [reply]
Was looking at the code again and am wondering, doesn't the fact that every rank loads world_size batches mean a lot of extra (unnecessary) work is being done? @pzelasko 
Wouldn't it be more efficient to have each sampler just load the batches for its rank?

**Reply by pzelasko** [reply]
It's a naive mechanism that ensures an equal number of steps in the loop over dataloader at each rank. Yes, it's `world_size - 1` times the overhead, but it's still negligible because we're only loading + parsing JSONL here. You'd only start noticing once you scale to thousands of ranks, at which point you would probably switch to a sequential I/O format such as Lhotse Shar anyway. With sequential I/O (tarred data) we use a different, more efficient mechanism to ensure the ranks never get desynchronized ([see the tutorial if interested](https://colab.research.google.com/github/lhotse-speech/lhotse/blob/master/examples/04-lhotse-shar.ipynb)).

**Reply by RuABraun** [reply]
yeah makes sense, the expensive operations are done when the cuts are passed to the dataset to be collated into a batch 👍
